frame/merkle-mountain-range/primitives/src/lib.rs

```diff
@@ -368,26 +373,58 @@ impl FullLeaf for OpaqueLeaf {
 	}
 }
 
+/// A type-safe wrapper for the concrete leaf type.
+///
+/// This structure serves merely to avoid passing raw `Vec<u8>` around.
+/// It must be `Vec<u8>`-encoding compatible.
+///
+/// It is different from [OpaqueLeaf], because it does implement `Codec`
+/// and the encoding has to match raw `Vec<u8>` encoding.
+#[derive(codec::Encode, codec::Decode, RuntimeDebug, PartialEq, Eq)]
+pub struct EncodableOpaqueLeaf(pub Vec<u8>);
+
+impl EncodableOpaqueLeaf {
+	/// Convert a concrete leaf into encodable opaque version.
+	pub fn from_leaf<T: FullLeaf>(leaf: &T) -> Self {
```

Comment:
Shouldn't this throw some kind of error, if `T` doesn't meet the encoding requirements?

Reply:
Hmm, but there isn't really any requirements other than `FullLeaf` implementation. And the encoding is infallible.

So `EncodableOpqaueLeaf` and `OpaqueLeaf` serve a bit different purposes.
Both are meant for leaves that we don't necessarily know the concrete type of (i.e. we know it's a leaf, but we don't have the struct/type to encode it anywhere in the code).

The first one is a wrapper type to send such "unknown leaves" to the RuntimeApi or over RPC.
The second one is for MMR implementation - i.e. it implements `FullLeaf`, but can't implement `Encode/Decode` because of conflicting trait implementations.

`from_leaf` is just a convenience method to convert a known leaf type into its encodable and opaque form - so there is no extra requirements here.